### PR TITLE
LG-2530 LG-2531 Provider inherits unset values. Exposes darkMode setter

### PR DESCRIPTION
**Pull request by TheSonOfThomp**
## ✍️ Proposed changes

🎟 _Jira ticket:_ [Nested provides](https://jira.mongodb.org/browse/LG-2530)
Nested LeafyGreenProviders now inherit values from ancestor contexts.
For example: 
```jsx
<LeafyGreenProvider darkMode={true}>
    Will have `darkMode == true` and `baseFontSize == 13`
  <LeafyGreenProvider baseFontSize={16}>
    Will have `darkMode == true` and `baseFontSize == 16`
  </LeafyGreenProvider>
</LeafyGreenProvider>,
```

🎟 _Jira ticket:_ [Dark Mode setter](https://jira.mongodb.org/browse/LG-2531)
Additionally we expose a `setDarkMode` function form the `useDarkMode` hook.

Note that the value can still be controlled externally by the prop on LeafyGreenProvider. The value of the last update is used as the context state.

```jsx
<LeafyGreenProvider darkMode={true}>
  <MyComponent />
</LeafyGreenProvider>

...

const MyComponent = () => {
  const { setDarkMode } = useDarkMode()
  ... 

}
```

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Style change (purely visual updates; if this is the case, attach a screenshot below!)

## ✅ Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

## Feedback Levels

Please try to mark your comments with one of these levels

1. Praise (👏 🚀 🔥 🍾)
2. Opinions, Ideas, Suggestions (💡 👀)
3. Let’s discuss @ Code review (🤷 🤔 🧐 )
4. Doesn’t meet agreed upon standards (😬 ⚠️ ‼️)
5. Missing an edge case (🛑 🙅 ❌)
6. Missing core functionality (⛔️ ☣️ 🚧 🌋 💀)

(Note: Use ⌘-⌃-␣ to open the Emoji keyboard!)

### CODE REVIEW

File: packages/leafygreen-provider/src/LeafyGreenContext.tsx [hunk 1]
```diff
@@ -18,15 +22,41 @@ type LeafyGreenProviderProps = {
...
+  const inheritedContainer = usePopoverPortalContainer();
+  const popoverPortalContainer =
+    popoverPortalContainerProp ?? inheritedContainer;
+
   return (
     <UsingKeyboardProvider>
       <PortalContextProvider popover={popoverPortalContainer}>
         <TypographyProvider baseFontSize={baseFontSize}>
-          <DarkModeProvider globalDarkMode={darkMode}>
+          <DarkModeProvider
```
Comment (shaneeza) [inline packages/leafygreen-provider/src/LeafyGreenContext.tsx:56]: 🤯

File: .changeset/stupid-flies-taste.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
+'@leafygreen-ui/leafygreen-provider': major
+---
+
+Nested LeafyGreenProviders now inherit values from ancestor contexts.
+For example:
+```jsx
+<LeafyGreenProvider darkMode={true}>
+    Will have `darkMode == true` and `baseFontSize == 13`
```
Comment (shaneeza) [inline .changeset/stupid-flies-taste.md:9]: This should be 14 until we update to 13?

File: packages/leafygreen-provider/src/LeafyGreenContext.tsx [hunk 3]
```diff
@@ -18,15 +22,41 @@ type LeafyGreenProviderProps = {
...
+  popoverPortalContainer: popoverPortalContainerProp,
+  darkMode: darkModeProp,
 }: PropsWithChildren<LeafyGreenProviderProps>) {
+  // If a darkMode prop is not set,
+  // then we want to check if there's an outer dark mode context.
+
+  // if the prop is set, we use that
+  // if the prop is not set, we use outer context
+  const { contextDarkMode: inheritedDarkMode } = useDarkModeContext();
+  // const darkMode = darkModeProp ?? inheritedDarkMode;
```
Comment (shaneeza) [inline packages/leafygreen-provider/src/LeafyGreenContext.tsx:35]: can we remove this?

File: stories/ProviderDemo.story.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import Card from '@leafygreen-ui/card';
+import { Body } from '@leafygreen-ui/typography';
+
+export default {
+  title: 'LeafyGreen Provider/DarkMode',
+};
+
+const DarkModeComponent = () => {
+  const { darkMode, theme, setDarkMode } = useDarkMode();
+  // console.log({ darkMode, theme });
```
Comment (shaneeza) [inline stories/ProviderDemo.story.tsx:16]: can we remove this?

File: packages/leafygreen-provider/src/LeafyGreenContext.tsx [hunk 5]
```diff
@@ -18,15 +22,41 @@ type LeafyGreenProviderProps = {
...
-  baseFontSize,
-  popoverPortalContainer,
-  darkMode,
+  baseFontSize: fontSizeProp,
+  popoverPortalContainer: popoverPortalContainerProp,
+  darkMode: darkModeProp,
 }: PropsWithChildren<LeafyGreenProviderProps>) {
+  // If a darkMode prop is not set,
+  // then we want to check if there's an outer dark mode context.
+
```
Comment (shaneeza) [inline packages/leafygreen-provider/src/LeafyGreenContext.tsx:31]: Seems like these comments are duplicates?

File: .changeset/tiny-vans-watch.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
+---
+'@leafygreen-ui/leafygreen-provider': minor
+---
+
+Exposes a `setDarkMode` function form the `useDarkMode` hook.
+Note that the value can still be controlled externally by the prop on LeafyGreenProvider. The value of the last update is used as the context state.
```
Comment (shaneeza) [inline .changeset/tiny-vans-watch.md:6]: The last sentence is a little unclear, do you mean that if you use `setDarkMode` then that will be used as the context state?

File: packages/leafygreen-provider/src/LeafyGreenContext.spec.tsx [hunk 7]
```diff
@@ -1,9 +1,35 @@
...
+  const baseFontSize = useBaseFontSize();
+  const { portalContainer, scrollContainer } = usePopoverPortalContainer();
+
+  return (
+    <>
+      <div data-testid="dark-mode" data-value={darkMode} />
+      <div data-testid="font-size" data-value={baseFontSize} />
+      <div
+        data-testid="portal-container"
+        id={portalContainer!.getAttribute('id') ?? ''}
```
Comment (shaneeza) [inline packages/leafygreen-provider/src/LeafyGreenContext.spec.tsx:23]: interesting approach 👍🏽

**Review by shaneeza — APPROVED** [review]
Very cool! I played around and everything seems to work as expected 🤞🏽. I left some minor non blocking comments.

File: .changeset/tiny-vans-watch.md (hunk 6 above)
Comment (TheSonOfThomp) [reply .changeset/tiny-vans-watch.md]: @shaneeza Updated the phrasing. Does it make more sense now?

File: .changeset/tiny-vans-watch.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+</LeafyGreenProvider>
+...
+
+const MyComponent = () => {
+  const { setDarkMode } = useDarkMode()
+  ... 
+}
+```
+
+Note: If the value of `darkMode` passed into LeafyGreenProvider changes between renders (e.g. from an external `useState` call) then this new value will be used until the next `setDarkMode` call.
```
Comment (shaneeza) [inline .changeset/tiny-vans-watch.md:19]: 👍🏽

**Pull request merged**